src/parser/mod.rs

```diff
@@ -249,40 +233,73 @@ impl Parser {
         preceded!(tag!("#"), take_until_and_consume!("\n"))
     );
 
-    method!(trs<Parser, CompleteStr, TRS<Var, Op>>, mut self,
+    method!(trs<Parser, CompleteStr, (Signature, TRS)>, mut self,
             add_return_error!(
                 ErrorKind::Custom(1),
                 do_parse!(
-                    rules: many0!(
+                    sigrules: many0!(
                         map!(
                             do_parse!(
                                 many0!(ws!(call_m!(self.comment))) >>
                                 rule: call_m!(self.rule) >>
                                 ws!(semicolon) >>
                                 many0!(ws!(call_m!(self.comment))) >>
                                 (rule)),
-                            |r| {self.clear_variables(); r})) >>
-                    (TRS::new(rules))))
+                            |r| {
+                                let sr = (self.sig.clone(), r);
+                                self.clear_variables();
+                                sr
+                            })) >>
+                    ({
+                        let (mut sig, _) = Signature::new(self.sig.operators.clone());
+                        let mut rules = Vec::new();
+                        for (rsig, rule) in sigrules {
+                            // we are using the same operators across rules, so it's safe to give
+                            // `true` for `same_ops`.
+                            let sigchange = sig.consolidate(rsig, true);
+                            let rule = sigchange.reify_rule(rule);
+                            rules.push(rule);
+                        }
+                        (sig, TRS::new(rules))
+                    })))
```

Comment:
Hmmm... this and the `program` parser have gotten much longer. I see what they're both doing. I wonder if there's a nice way to refactor, though, or if this is the right place for all the `Signature` consolidation?